lib/fortnox/api/repositories/base/loaders.rb

```diff
@@ -7,8 +7,15 @@ module API
     module Repository
       module Loaders
         def all
-          response_hash = get(self.class::URI)
-          instantiate_collection_response(response_hash)
+          results = []
+
+          loop do
```

Comment:
Something like this might be a little clearer:
```ruby
# In the pagination module
def pages
  (1..total_pages ).to_a # Returns an array with [1,2,...,n] where n == total_pages
end

# In this file
def all
  pages.each_with_object([]) do |page_number, results|
    response_hash = get(self.class::URI, query: { page: page_number })
    results.push( *instantiate_collection_response( response_hash ))
  end
end
```
That way we don't have to do bounds checking, so we can drop `last_page?` and get the page number for free, so we can drop `next_page`. And we get a cleaner method without temporary variables 😄

Reply:
Hmm.. I like that way that looks but that doesn't solve for the inital request to get the number of total pages.. I guess we could do one request and then only start paginating if there are more pages but that makes it kind of ugly.. Any suggestions?

Reply:
My updated suggestion:
```ruby
def all( limit: Float::INFINITY )
  # Cast and clamp to natural number
  requested_number_of_pages = Integer(limit) < 1 ? 1 : Integer(limit)
  # Max 500 entities per request
  request_limit = [requested_number_of_pages, 500].max
  # Start at the first page
  page = 1
  results = []

  loop do
    response_hash = get(self.class::URI, query: { page: page, limit: request_limit })
    results.push( *instantiate_collection_response( response_hash ))

    limit_reached = result.length >= requested_number_of_pages
    no_more_pages = @metadata.total_pages == page
    break if limit_reached or no_more_pages
    page += 1
  end

  results = results(0...requested_number_of_pages) unless requested_number_of_pages.infinite?

  return results
end
```

Reply:
[I updated it a bit after I ran it locally and got some errors]

That code works for getting all when < 500 as well as limiting the result to the requested limit. I tested pagination by setting the `request_limit` to 10 and making sure the query still returned the right number of entities.

The test code I used, after starting the debug console with `bin/console`, was this:
```ruby
reload!
Fortnox::API.configure do |config|
  config.client_secret = '9aBA8ZgsvR'
  config.access_token = 'ccaef817-d5d8-4b1c-a316-54f3e55c5c54'
end
r = Fortnox::API::Repository::Invoice.new
r.all.length
```
Sometimes with a `limit:` argument to the all call.

Reply:
[I updated it a bit after I ran it locally and got some errors]

That code works for getting all when < 500 as well as limiting the result to the requested limit. I tested pagination by setting the `request_limit` to 10 and making sure the query still returned the right number of entities.

The test code I used, after starting the debug console with `bin/console`, was this:
```ruby
reload!
Fortnox::API.configure do |config|
  config.client_secret = '[SECRET]'
  config.access_token = '[UUID]'
end
r = Fortnox::API::Repository::Invoice.new
r.all.length
```
Sometimes with a `limit:` argument to the all call.